Change Transport/Configurer classes to require interface.

DavidBoike opened the pull request:
Related to #3203 

Replaced the invocation of Configure{Transport/Persistence} classes as dynamic methods and reflection with a full interface `IConfigureTestEnvironment`. Saw a lot of duplication between transport and persistence and eliminated that.

Stopped short of replacing the settings as string dictionary with a class having defined properties, since this concept touches builders and serializers in addition to transports and persistences. Also didn't apply the requirement for an interface-implementing configurer class on builders/serializers.

If settings was replaced, the list of properties would (I think) be:

* UniqueTestRunId - Guid I would add that would uniquely identify each test run. Azure tests need this kind of concept to enable ignoring messages in queue from any other test run. That behavior could potentially be generalized.
* Transport
* TransportConnectionString
* Persistence
* PersistenceConnectionString
* Builder
* Serializer

Thoughts? @SeanFeldman @danielmarbach @andreasohlund @SimonCropp @Particular/nservicebus

LiamFeldman commented on src/NServiceBus.AcceptanceTesting/NServiceBus.AcceptanceTesting.csproj:
```diff
@@ -66,6 +66,7 @@
     <Compile Include="Support\EndpointConfiguration.cs" />
     <Compile Include="Support\EndpointRunner.cs" />
     <Compile Include="Support\FailTestOnErrorMessageFeature.cs" />
+    <Compile Include="Support\IConfigureTestEnvironment.cs" />
```
Probably nitpicking here, but doesn't this allow to configure test run rather than environment?
`IConfigureTestRun` or `IConfigureTestExecution`

SeanFeldman replied:
And my son didn't log out from my machine :)

andreasohlund replied:
>And my son didn't log out from my machine :)

Who cares, he's got a GH account! Is he looking for work?

SeanFeldman commented on src/NServiceBus.AcceptanceTesting/Support/IConfigureTestEnvironment.cs:
```diff
@@ -0,0 +1,12 @@
+
+    public interface IConfigureTestEnvironment
+    {
+        Task Configure(BusConfiguration configuration, IDictionary<string, string> settings);
```
What if we rename this to `Initialize` since that's what it does?

SeanFeldman commented on src/NServiceBus.AcceptanceTesting/Support/IConfigureTestEnvironment.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        Task Configure(BusConfiguration configuration, IDictionary<string, string> settings);
+
+        Task Cleanup();
```
And this method would become `Configure` or `Execute`?

SeanFeldman commented on src/NServiceBus.AcceptanceTesting/Support/EndpointRunner.cs:
```diff
@@ -173,13 +173,13 @@ void RegisterInheritanceHierarchyOfContextInSettings(ScenarioContext context)
 
         async Task Cleanup()
```
To be good citizens, should this method name change to what it stands for?

andreasohlund replied:
Isn't `Cleanup` what it does?

SeanFeldman replied:
My bad. Please ignore.

SeanFeldman commented on src/NServiceBus.AcceptanceTests/ConfigureMsmqTransport.cs:
```diff
@@ -4,10 +4,11 @@
 using NServiceBus.Transports;
 
-public class ConfigureMsmqTransport
+public class ConfigureMsmqTransport : IConfigureTestEnvironment
```
:+1:

SeanFeldman commented on src/NServiceBus.AcceptanceTesting/Support/IConfigureTestEnvironment.cs:
```diff
@@ -0,0 +1,12 @@
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IConfigureTestEnvironment
```
I know that xml comments are not popular, yet would suggest we put a description on this (what's the purpose of this contract).

SeanFeldman commented:
@DavidBoike didn't see any trace of `UniqueTestRunId`. Is it something that we should add as a separate PR?

DavidBoike commented:
Not sure it makes sense to add the test run id unless explicit properties are used. In Azure we can still create our own test run id in the dictionary.

SeanFeldman commented:
Ok

danielmarbach commented on src/NServiceBus.AcceptanceTests/EndpointTemplates/ConfigureExtensions.cs:
```diff
@@ -63,38 +37,43 @@ public static void DefineTransactions(this BusConfiguration config, IDictionary<
-            {
-                throw new InvalidOperationException($"Class {configurerType.FullName} must include a \"public Task Configure(BusConfiguration configuration, IDictionary<string, string> settings)\" method.", rbe);
-            }
+            await configurer.Configure(config, settings);
```
`ConfigureAwait(false)` since it is "framework"

DavidBoike commented:
Renamed IConfigureTestEnvironment to IConfigureTestExecution, added XML comments on the interface, and added ConfigureAwait(false). Please review again?

SeanFeldman commented:
@DavidBoike strange, one test is failing. Is that a flaky one?

@danielmarbach raised another issue that directly ties into this one: https://github.com/Particular/NServiceBus/issues/3223

DavidBoike commented:
I'm going to add a `UniqueTestRunId` on the ScenarioContext class quick. I think it belongs there.

andreasohlund commented on src/NServiceBus.AcceptanceTesting/ScenarioContext.cs:
```diff
@@ -7,6 +7,8 @@
 
     public abstract class ScenarioContext
     {
+        public Guid UniqueTestRunId { get; } = Guid.NewGuid();
```
Nitpick: Not sure what the "Unique" prefix adds, `public Guid TestRunId` seems pretty revealing?

andreasohlund commented on src/NServiceBus.AcceptanceTesting/Support/EndpointRunner.cs:
```diff
@@ -173,13 +173,13 @@ void RegisterInheritanceHierarchyOfContextInSettings(ScenarioContext context)
+            IConfigureTestExecution transportCleaner;
             if (busConfiguration.GetSettings().TryGet("CleanupTransport", out transportCleaner))
             {
                 await transportCleaner.Cleanup();
```
Since this is technically infra should we do `.ConfigureAwait(false)`?

danielmarbach replied:
Deffo!

DavidBoike replied:
Yes, how did I miss that one!?

andreasohlund replied:
The amount of :beer: owed to @danielmarbach in Thailand is getting worrying, I might have to chip in and help him drink it! :)

andreasohlund commented on src/NServiceBus.AcceptanceTesting/Support/EndpointRunner.cs:
```diff
@@ -173,13 +173,13 @@ void RegisterInheritanceHierarchyOfContextInSettings(ScenarioContext context)
+            IConfigureTestExecution persistenceCleaner;
             if (busConfiguration.GetSettings().TryGet("CleanupPersistence", out persistenceCleaner))
             {
                 await persistenceCleaner.Cleanup();
```
Since this is technically infra should we do `.ConfigureAwait(false)`?

danielmarbach replied:
Deffo!

andreasohlund commented on src/NServiceBus.AcceptanceTesting/Support/EndpointRunner.cs:
```diff
@@ -173,13 +173,13 @@ void RegisterInheritanceHierarchyOfContextInSettings(ScenarioContext context)
 
-            dynamic persistenceCleaner;
+            IConfigureTestExecution persistenceCleaner;
             if (busConfiguration.GetSettings().TryGet("CleanupPersistence", out persistenceCleaner))
```
Do we need to separate them like this or could we just have a 
```
List<IConfigureTestExecution> cleaners;
busConfiguration.GetSettings().TryGet(out cleaners)
```
instead?

That way we can just add more in the future without touching this code? (or does the order matter in some way)

DavidBoike replied:
This could potentially expand to builders/serializers too. Don't think execution order matters. I'll make the change.

andreasohlund commented on src/NServiceBus.AcceptanceTests/EndpointTemplates/ConfigureExtensions.cs:
```diff
@@ -63,38 +37,43 @@ public static void DefineTransactions(this BusConfiguration config, IDictionary<
+            var dependencyType = Type.GetType(settings[dependencyTypeString]);
 
-            var typeName = "Configure" + persistenceType.Name;
+            var typeName = "Configure" + dependencyType.Name;
```
With my above "List<>" suggestion we could make this just look for loaded types assignable to `IConfigureTestExecution` ?

danielmarbach replied:
Do we really need that list or is it YAGNI?

andreasohlund replied:
Since it means less code and less stringly typed stuff I'd vote to go ahead

DavidBoike commented:
I made the cleaners a list and invoked them asynchronously w/ Task.WaitAll(). Hopefully that could gain some small test execution speedups.

I'm gonna agree with YAGNI on loading types assignable to IConfigureTestExecution for now. That could be introduced later but I need the changes to the public API surface now.

Removed the Unique from TestRunId and fixed the ConfigureAwaits.

Tests all passing.

danielmarbach commented:
Luckily you used WhenAll instead of WaitAll :)
